File: ctx-pack/ctx_pack.py
#!/usr/bin/env python3
"""ctx_pack.py — shrink logs/CI output/dumps before pasting them into an LLM.

Reads text, removes what a model does not need (ANSI codes, repeated lines,
giant base64 blobs, boilerplate), optionally keeps only error lines with
context, and prints the packed result to stdout. A before/after size report
(bytes, lines, estimated tokens) goes to stderr so the savings are visible.

Python 3.9+, no dependencies. Token estimate is chars/4 — rough, but fine
for "did this get 10x smaller?".

Usage:
  ./ctx_pack.py build.log                     # pack a file to stdout
  some-command 2>&1 | ./ctx_pack.py           # pack a pipe
  ./ctx_pack.py --errors-only -C 5 ci.log     # only error lines ±5 context
  ./ctx_pack.py --max-tokens 8000 huge.log    # hard budget: keeps head+tail
  ./ctx_pack.py --keep 'my_service' app.log   # never drop matching lines

Options:
  FILE ...           input files (default: stdin); multiple files are packed
                     in sequence with a "==> name <==" header each
  --errors-only      keep only lines matching the error pattern (plus context)
  -C, --context N    context lines around kept lines in --errors-only (default 3)
  --errors-regex RE  override the built-in error pattern
  --keep RE          lines matching RE are always kept verbatim (repeatable)
  --max-line N       truncate lines longer than N chars (default 400, 0=off)
  --max-tokens N     if the packed result still exceeds ~N tokens, keep the
                     first and last halves and elide the middle (0=off)
  --no-dedupe        don't collapse consecutive (near-)duplicate lines
  --no-blobs         don't squash long base64/hex runs
  --quiet            suppress the stats report on stderr

Exit codes: 0 ok, 1 bad args / unreadable input.
"""

import argparse
import re
import sys

ANSI_RE = re.compile(r"\x1b\[[0-9;?]*[a-zA-Z]|\x1b\][^\x07]*(\x07|\x1b\\)")
# unbroken base64ish / hex / hash runs that carry no meaning for a model
BLOB_RE = re.compile(r"[A-Za-z0-9+/_-]{64,}={0,2}")
# used to decide two lines are "the same" modulo counters/ids/timestamps
NORM_SUBS = [
    (re.compile(r"\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:?\d{2})?"), "<ts>"),
    (re.compile(r"\b0x[0-9a-fA-F]+\b"), "<hex>"),
    (re.compile(r"\b[0-9a-fA-F]{8,}\b"), "<hex>"),
    (re.compile(r"\b\d+\b"), "<n>"),
]
DEFAULT_ERRORS_RE = (
    r"(?i)\b(error|err!|fail(ed|ure)?|exception|traceback|fatal|panic"
    r"|assert(ion)?|segfault|denied|refused|timeout|timed out|unhandled"
    r"|cannot|not found|undefined|null pointer|stack trace)\b"
)


def estimate_tokens(text):
    return len(text) // 4


def normalize(line):
    for rx, repl in NORM_SUBS:
        line = rx.sub(repl, line)
    return line


def pack_lines(lines, args, keep_res):
    out = []

    def emit(line):
        if args.max_line and len(line) > args.max_line:
            line = line[: args.max_line] + f" …[+{len(line) - args.max_line} chars]"
        out.append(line)

    # 1. strip ANSI, trailing whitespace; squash blobs
    cleaned = []
    for line in lines:
        line = ANSI_RE.sub("", line).rstrip()
        if not args.no_blobs:
            line = BLOB_RE.sub(lambda m: f"<blob:{len(m.group(0))} chars>", line)
        cleaned.append(line)

    # 2. --errors-only: keep matching lines ± context, plus --keep lines
    if args.errors_only:
        err_re = re.compile(args.errors_regex)
        keep_idx = set()
        for i, line in enumerate(cleaned):
            if err_re.search(line) or any(r.search(line) for r in keep_res):
                keep_idx.update(range(max(0, i - args.context),
                                      min(len(cleaned), i + args.context + 1)))
        selected, last = [], None
        for i in sorted(keep_idx):
            if last is not None and i > last + 1:
                selected.append(f"⋮ [{i - last - 1} lines skipped]")
            selected.append(cleaned[i])
            last = i
        if last is not None and last < len(cleaned) - 1:
            selected.append(f"⋮ [{len(cleaned) - 1 - last} lines skipped]")
        cleaned = selected

    # 3. collapse blank runs and consecutive near-duplicates
    prev_norm, run_first, run_count, blank_run = None, None, 0, 0

    def flush_run():
        nonlocal run_first, run_count, prev_norm
        if run_first is not None:
            emit(run_first)
            if run_count > 1:
                emit(f"  [previous line repeated {run_count - 1} more times]")
        run_first, run_count, prev_norm = None, 0, None

    for line in cleaned:
        if not line:
            flush_run()
            blank_run += 1
            if blank_run == 1:
                out.append("")
            continue
        blank_run = 0
        if args.no_dedupe or any(r.search(line) for r in keep_res):
            flush_run()
            emit(line)
            continue
        norm = normalize(line)
        if norm == prev_norm:
            run_count += 1
        else:
            flush_run()
            run_first, run_count, prev_norm = line, 1, norm
    flush_run()

    # 4. hard token budget: keep head + tail
    if args.max_tokens:
        text = "\n".join(out)
        if estimate_tokens(text) > args.max_tokens:
            budget_chars = args.max_tokens * 4
            head_chars = budget_chars * 6 // 10   # errors usually cluster at the end,
            tail_chars = budget_chars - head_chars  # but the head sets the scene
            head, tail, dropped = [], [], 0
            n = 0
            for line in out:
                if n + len(line) <= head_chars:
                    head.append(line)
                    n += len(line) + 1
                else:
                    break
            n = 0
            for line in reversed(out[len(head):]):
                if n + len(line) <= tail_chars:
                    tail.append(line)
                    n += len(line) + 1
                else:
                    break
            dropped = len(out) - len(head) - len(tail)
            if dropped > 0:
                out = head + [f"⋮ [{dropped} lines elided to fit "
                              f"~{args.max_tokens} token budget]"] + list(reversed(tail))

    return out


def main():
    ap = argparse.ArgumentParser(
        description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("files", nargs="*")
    ap.add_argument("--errors-only", action="store_true")
    ap.add_argument("-C", "--context", type=int, default=3)
    ap.add_argument("--errors-regex", default=DEFAULT_ERRORS_RE)
    ap.add_argument("--keep", action="append", default=[])
    ap.add_argument("--max-line", type=int, default=400)
    ap.add_argument("--max-tokens", type=int, default=0)
    ap.add_argument("--no-dedupe", action="store_true")
    ap.add_argument("--no-blobs", action="store_true")
    ap.add_argument("--quiet", action="store_true")
    args = ap.parse_args()

    try:
        keep_res = [re.compile(p) for p in args.keep]
        re.compile(args.errors_regex)
    except re.error as e:
        sys.exit(f"error: bad regex: {e}")

    sources = []
    if args.files:
        for path in args.files:
            try:
                with open(path, encoding="utf-8", errors="replace") as fh:
                    sources.append((path, fh.read()))
            except OSError as e:
                sys.exit(f"error: {e}")
    else:
        sources.append(("stdin", sys.stdin.read()))

    in_chars = sum(len(t) for _, t in sources)
    in_lines = sum(t.count("\n") + 1 for _, t in sources)

    chunks = []
    for name, text in sources:
        packed = pack_lines(text.splitlines(), args, keep_res)
        if len(sources) > 1:
            chunks.append(f"==> {name} <==")
        chunks.extend(packed)
    result = "\n".join(chunks)
    sys.stdout.write(result + ("\n" if result else ""))

    if not args.quiet:
        out_chars = len(result)
        saved = 100.0 * (1 - out_chars / in_chars) if in_chars else 0.0
        print(
            f"ctx-pack: {in_chars:,} chars / {in_lines:,} lines "
            f"(~{in_chars // 4:,} tok) -> "
            f"{out_chars:,} chars / {len(chunks):,} lines "
            f"(~{estimate_tokens(result):,} tok)  saved {saved:.1f}%",
            file=sys.stderr,
        )


if __name__ == "__main__":
    main()

File: ctx-pack/ctx_pack.ts
#!/usr/bin/env -S npx tsx
/**
 * ctx_pack.ts — shrink logs/CI output/dumps before pasting them into an LLM.
 *
 * TypeScript port of ctx_pack.py (functionally equivalent). Reads text, removes
 * what a model does not need (ANSI codes, repeated lines, giant base64 blobs,
 * boilerplate), optionally keeps only error lines with context, and prints the
 * packed result to stdout. A before/after size report (chars, lines, estimated
 * tokens) goes to stderr so the savings are visible.
 *
 * Node >= 22, zero runtime deps (stdlib only), ESM, run via `npx tsx` (no build).
 * Token estimate is chars/4 — rough, but fine for "did this get 10x smaller?".
 *
 * Usage:
 *   ./ctx_pack.ts build.log                     # pack a file to stdout
 *   some-command 2>&1 | ./ctx_pack.ts           # pack a pipe
 *   ./ctx_pack.ts --errors-only -C 5 ci.log     # only error lines +/-5 context
 *   ./ctx_pack.ts --max-tokens 8000 huge.log    # hard budget: keeps head+tail
 *   ./ctx_pack.ts --keep 'my_service' app.log   # never drop matching lines
 *
 * Options:
 *   FILE ...           input files (default: stdin); multiple files are packed
 *                      in sequence with a "==> name <==" header each
 *   --errors-only      keep only lines matching the error pattern (plus context)
 *   -C, --context N    context lines around kept lines in --errors-only (def 3)
 *   --errors-regex RE  override the built-in error pattern
 *   --keep RE          lines matching RE are always kept verbatim (repeatable)
 *   --max-line N       truncate lines longer than N chars (default 400, 0=off)
 *   --max-tokens N     if the packed result still exceeds ~N tokens, keep the
 *                      first and last halves and elide the middle (0=off)
 *   --no-dedupe        don't collapse consecutive (near-)duplicate lines
 *   --no-blobs         don't squash long base64/hex runs
 *   --quiet            suppress the stats report on stderr
 *   -h, --help         print this usage and exit non-zero
 *
 * Exit codes: 0 ok, 1 bad args / unreadable input, 2 usage/help.
 */

import { readFileSync } from "node:fs";
import process from "node:process";
import { fileURLToPath } from "node:url";

const ANSI_RE = /\x1b\[[0-9;?]*[a-zA-Z]|\x1b\][^\x07]*(\x07|\x1b\\)/g;
// unbroken base64ish / hex / hash runs that carry no meaning for a model
const BLOB_RE = /[A-Za-z0-9+/_-]{64,}={0,2}/g;
// used to decide two lines are "the same" modulo counters/ids/timestamps
const NORM_SUBS: [RegExp, string][] = [
  [/\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:?\d{2})?/g, "<ts>"],
  [/\b0x[0-9a-fA-F]+\b/g, "<hex>"],
  [/\b[0-9a-fA-F]{8,}\b/g, "<hex>"],
  [/\b\d+\b/g, "<n>"],
];
const DEFAULT_ERRORS_RE =
  "\\b(error|err!|fail(ed|ure)?|exception|traceback|fatal|panic" +
  "|assert(ion)?|segfault|denied|refused|timeout|timed out|unhandled" +
  "|cannot|not found|undefined|null pointer|stack trace)\\b";

export interface Options {
  files: string[];
  errorsOnly: boolean;
  context: number;
  errorsRegex: string;
  keep: string[];
  maxLine: number;
  maxTokens: number;
  noDedupe: boolean;
  noBlobs: boolean;
  quiet: boolean;
}

export interface PackResult {
  stdout: string;
  stderr: string;
  code: number;
}

function defaultOptions(): Options {
  return {
    files: [],
    errorsOnly: false,
    context: 3,
    errorsRegex: DEFAULT_ERRORS_RE,
    keep: [],
    maxLine: 400,
    maxTokens: 0,
    noDedupe: false,
    noBlobs: false,
    quiet: false,
  };
}

export function estimateTokens(text: string): number {
  return Math.floor(text.length / 4);
}

export function normalize(line: string): string {
  for (const [rx, repl] of NORM_SUBS) {
    line = line.replace(rx, repl);
  }
  return line;
}

/** Mirror of Python str.splitlines() for the common line terminators. */
function splitLines(text: string): string[] {
  if (text === "") return [];
  const parts = text.split(/\r\n|\r|\n/);
  // Python's splitlines() does not emit a trailing "" for a final terminator.
  if (/[\r\n]$/.test(text)) parts.pop();
  return parts;
}

function fmt(n: number): string {
  return n.toLocaleString("en-US");
}

export function packLines(
  lines: string[],
  opts: Options,
  keepRes: RegExp[],
): string[] {
  const out: string[] = [];

  const emit = (line: string): void => {
    if (opts.maxLine && line.length > opts.maxLine) {
      line =
        line.slice(0, opts.maxLine) +
        ` …[+${line.length - opts.maxLine} chars]`;
    }
    out.push(line);
  };

  // 1. strip ANSI, trailing whitespace; squash blobs
  let cleaned: string[] = [];
  for (let line of lines) {
    line = line.replace(ANSI_RE, "").replace(/\s+$/, "");
    if (!opts.noBlobs) {
      line = line.replace(BLOB_RE, (m) => `<blob:${m.length} chars>`);
    }
    cleaned.push(line);
  }

  // 2. --errors-only: keep matching lines +/- context, plus --keep lines
  if (opts.errorsOnly) {
    const errRe = new RegExp(opts.errorsRegex, "i");
    const keepIdx = new Set<number>();
    for (let i = 0; i < cleaned.length; i++) {
      const line = cleaned[i];
      if (errRe.test(line) || keepRes.some((r) => r.test(line))) {
        const lo = Math.max(0, i - opts.context);
        const hi = Math.min(cleaned.length, i + opts.context + 1);
        for (let j = lo; j < hi; j++) keepIdx.add(j);
      }
    }
    const selected: string[] = [];
    let last: number | null = null;
    for (const i of [...keepIdx].sort((a, b) => a - b)) {
      if (last !== null && i > last + 1) {
        selected.push(`⋮ [${i - last - 1} lines skipped]`);
      }
      selected.push(cleaned[i]);
      last = i;
    }
    if (last !== null && last < cleaned.length - 1) {
      selected.push(`⋮ [${cleaned.length - 1 - last} lines skipped]`);
    }
    cleaned = selected;
  }

  // 3. collapse blank runs and consecutive near-duplicates
  let prevNorm: string | null = null;
  let runFirst: string | null = null;
  let runCount = 0;
  let blankRun = 0;

  const flushRun = (): void => {
    if (runFirst !== null) {
      emit(runFirst);
      if (runCount > 1) {
        emit(`  [previous line repeated ${runCount - 1} more times]`);
      }
    }
    runFirst = null;
    runCount = 0;
    prevNorm = null;
  };

  for (const line of cleaned) {
    if (!line) {
      flushRun();
      blankRun += 1;
      if (blankRun === 1) out.push("");
      continue;
    }
    blankRun = 0;
    if (opts.noDedupe || keepRes.some((r) => r.test(line))) {
      flushRun();
      emit(line);
      continue;
    }
    const norm = normalize(line);
    if (norm === prevNorm) {
      runCount += 1;
    } else {
      flushRun();
      runFirst = line;
      runCount = 1;
      prevNorm = norm;
    }
  }
  flushRun();

  // 4. hard token budget: keep head + tail
  if (opts.maxTokens) {
    const text = out.join("\n");
    if (estimateTokens(text) > opts.maxTokens) {
      const budgetChars = opts.maxTokens * 4;
      const headChars = Math.floor((budgetChars * 6) / 10);
      const tailChars = budgetChars - headChars;
      const head: string[] = [];
      const tail: string[] = [];
      let n = 0;
      for (const line of out) {
        if (n + line.length <= headChars) {
          head.push(line);
          n += line.length + 1;
        } else {
          break;
        }
      }
      n = 0;
      const rest = out.slice(head.length);
      for (let k = rest.length - 1; k >= 0; k--) {
        const line = rest[k];
        if (n + line.length <= tailChars) {
          tail.push(line);
          n += line.length + 1;
        } else {
          break;
        }
      }
      const dropped = out.length - head.length - tail.length;
      if (dropped > 0) {
        return [
          ...head,
          `⋮ [${dropped} lines elided to fit ~${opts.maxTokens} token budget]`,
          ...tail.reverse(),
        ];
      }
    }
  }

  return out;
}

const USAGE = `ctx-pack — shrink logs/CI output/dumps before pasting into an LLM.

Usage:
  ctx_pack.ts [options] [FILE ...]
  some-command 2>&1 | ctx_pack.ts [options]

Options:
  --errors-only      keep only lines matching the error pattern (plus context)
  -C, --context N    context lines around kept lines in --errors-only (default 3)
  --errors-regex RE  override the built-in error pattern
  --keep RE          lines matching RE are always kept verbatim (repeatable)
  --max-line N       truncate lines longer than N chars (default 400, 0=off)
  --max-tokens N     keep head+tail and elide the middle if over ~N tokens (0=off)
  --no-dedupe        don't collapse consecutive (near-)duplicate lines
  --no-blobs         don't squash long base64/hex runs
  --quiet            suppress the stats report on stderr
  -h, --help         print this usage and exit non-zero`;

interface ParseResult {
  opts?: Options;
  help?: boolean;
  error?: string;
}

function parseArgs(argv: string[]): ParseResult {
  const opts = defaultOptions();
  let onlyFiles = false;

  const parseInt10 = (s: string, flag: string): number | null => {
    if (!/^[+-]?\d+$/.test(s)) return null;
    const v = Number.parseInt(s, 10);
    return Number.isNaN(v) ? null : v;
  };

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (onlyFiles) {
      opts.files.push(a);
      continue;
    }
    const eq = a.indexOf("=");
    const flag = a.startsWith("--") && eq !== -1 ? a.slice(0, eq) : a;
    const inlineVal = a.startsWith("--") && eq !== -1 ? a.slice(eq + 1) : null;
    const takeVal = (): string | null => {
      if (inlineVal !== null) return inlineVal;
      if (i + 1 < argv.length) return argv[++i];
      return null;
    };

    if (a === "--") {
      onlyFiles = true;
    } else if (a === "-h" || a === "--help") {
      return { help: true };
    } else if (flag === "--errors-only") {
      opts.errorsOnly = true;
    } else if (flag === "--no-dedupe") {
      opts.noDedupe = true;
    } else if (flag === "--no-blobs") {
      opts.noBlobs = true;
    } else if (flag === "--quiet") {
      opts.quiet = true;
    } else if (flag === "-C" || flag === "--context") {
      const raw = flag === "-C" ? takeVal() : takeVal();
      if (raw === null) return { error: `argument ${flag}: expected one argument` };
      const v = parseInt10(raw, flag);
      if (v === null) return { error: `argument -C/--context: invalid int value: '${raw}'` };
      opts.context = v;
    } else if (a.startsWith("-C") && a.length > 2 && a[2] !== "-") {
      const raw = a.slice(2);
      const v = parseInt10(raw, "-C");
      if (v === null) return { error: `argument -C/--context: invalid int value: '${raw}'` };
      opts.context = v;
    } else if (flag === "--errors-regex") {
      const raw = takeVal();
      if (raw === null) return { error: "argument --errors-regex: expected one argument" };
      opts.errorsRegex = raw;
    } else if (flag === "--keep") {
      const raw = takeVal();
      if (raw === null) return { error: "argument --keep: expected one argument" };
      opts.keep.push(raw);
    } else if (flag === "--max-line") {
      const raw = takeVal();
      if (raw === null) return { error: "argument --max-line: expected one argument" };
      const v = parseInt10(raw, "--max-line");
      if (v === null) return { error: `argument --max-line: invalid int value: '${raw}'` };
      opts.maxLine = v;
    } else if (flag === "--max-tokens") {
      const raw = takeVal();
      if (raw === null) return { error: "argument --max-tokens: expected one argument" };
      const v = parseInt10(raw, "--max-tokens");
      if (v === null) return { error: `argument --max-tokens: invalid int value: '${raw}'` };
      opts.maxTokens = v;
    } else if (a.length > 1 && a.startsWith("-") && a !== "-") {
      return { error: `unrecognized arguments: ${a}` };
    } else {
      opts.files.push(a);
    }
  }

  return { opts };
}

/**
 * Run the packer with the given argv and stdin text, returning captured output
 * instead of touching process streams. Pure enough to unit-test directly.
 */
export function pack(argv: string[], stdin: string): PackResult {
  const parsed = parseArgs(argv);
  if (parsed.help) {
    return { stdout: USAGE + "\n", stderr: "", code: 2 };
  }
  if (parsed.error || !parsed.opts) {
    return { stdout: "", stderr: `error: ${parsed.error}\n`, code: 2 };
  }
  const opts = parsed.opts;

  let keepRes: RegExp[];
  try {
    keepRes = opts.keep.map((p) => new RegExp(p));
    new RegExp(opts.errorsRegex);
  } catch (e) {
    return { stdout: "", stderr: `error: bad regex: ${(e as Error).message}\n`, code: 1 };
  }

  const sources: [string, string][] = [];
  if (opts.files.length > 0) {
    for (const path of opts.files) {
      try {
        sources.push([path, readFileSync(path, "utf-8")]);
      } catch (e) {
        return { stdout: "", stderr: `error: ${(e as Error).message}\n`, code: 1 };
      }
    }
  } else {
    sources.push(["stdin", stdin]);
  }

  let inChars = 0;
  let inLines = 0;
  for (const [, t] of sources) {
    inChars += t.length;
    inLines += (t.split("\n").length - 1) + 1;
  }

  const chunks: string[] = [];
  for (const [name, text] of sources) {
    const packed = packLines(splitLines(text), opts, keepRes);
    if (sources.length > 1) chunks.push(`==> ${name} <==`);
    chunks.push(...packed);
  }
  const result = chunks.join("\n");
  const stdout = result + (result ? "\n" : "");

  let stderr = "";
  if (!opts.quiet) {
    const outChars = result.length;
    const saved = inChars ? 100.0 * (1 - outChars / inChars) : 0.0;
    stderr =
      `ctx-pack: ${fmt(inChars)} chars / ${fmt(inLines)} lines ` +
      `(~${fmt(Math.floor(inChars / 4))} tok) -> ` +
      `${fmt(outChars)} chars / ${fmt(chunks.length)} lines ` +
      `(~${fmt(estimateTokens(result))} tok)  saved ${saved.toFixed(1)}%\n`;
  }

  return { stdout, stderr, code: 0 };
}

function main(): void {
  const argv = process.argv.slice(2);
  const hasFiles = parseArgs(argv).opts?.files.length ?? 0;
  let stdin = "";
  if (hasFiles === 0) {
    try {
      stdin = readFileSync(0, "utf-8");
    } catch {
      stdin = "";
    }
  }
  const { stdout, stderr, code } = pack(argv, stdin);
  if (stdout) process.stdout.write(stdout);
  if (stderr) process.stderr.write(stderr);
  process.exit(code);
}

const isMain =
  process.argv[1] !== undefined &&
  fileURLToPath(import.meta.url) === process.argv[1];
if (isMain) {
  main();
}
